bfvmm/src/serial/src/serial_port_x86.cpp

```diff
@@ -0,0 +1,430 @@
+
+#include <serial_port_x86.h>
+
+struct __divisor_rate divisor_table[] =
+{
+    {50, 0x900},
+    {75, 0x600},
+    {110, 0x417},
+    {150, 0x300},
+    {300, 0x180},
+    {600, 0xC0},
+    {1200, 0x60},
+    {1800, 0x40},
+    {2000, 0x3A},
+    {2400, 0x30},
+    {3600, 0x20},
+    {4800, 0x18},
+    {7200, 0x10},
+    {9600, 0xC},
+    {19200, 0x6},
+    {38400, 0x3},
+    {57600, 0x2},
+    {115200, 0x1},
+    {0, 0},
+};
+
+struct __parity_bits parity_table[] =
+{
+    { NONE, 0x00},
+    { ODD,  0x08},
+    { EVEN, 0x18},
+    { MARK, 0x28},
+    { SPACE, 0x38},
+    { __PARITY_MAX, 0x00 },
+};
+
+struct __data_bits data_size_table[] =
+{
+    { 5, 0x00 },
+    { 6, 0x01 },
+    { 7, 0x02 },
+    { 8, 0x03 },
+    { 0, 0x00 },
+};
+
+serial_port_x86::serial_port_x86(port_io &io, uint8_t port, uint32_t baud, uint8_t data_size, PARITY_MODE parity, uint8_t stop_bits) : m_port_io(io)
+{
+    switch (port)
+    {
+        case 1:
+        {
+            m_port = COM1_IO_PORT;
+            break;
+        }
+        case 2:
+        {
+            m_port = COM2_IO_PORT;
+            break;
+        }
+        case 3:
+        {
+            m_port = COM3_IO_PORT;
+            break;
+        }
+        case 4:
+        {
+            m_port = COM4_IO_PORT;
+            break;
+        }
+        default:
+        {
+            // Error case, not a valid port
+            m_port = 0xFFFF;
+            return;
+        }
+    }
+
+    m_baud = baud;
+    m_parity = parity;
+    m_data_size = data_size;
+    m_stop_bits = stop_bits;
+}
+
+serial_port_x86::~serial_port_x86(void)
+{
+
+}
+
+serial::errno
+serial_port_x86::open(void)
+{
+    disable_interrupt_mode();
+
+    set_baud_rate(m_baud);
+    set_parity_mode(m_parity);
+    set_data_size(m_data_size);
+    set_stop_bits(m_stop_bits);
+    enable_fifo();
+
+    enable_interrupt_mode(m_interrupt_mode);
+
+    return serial::SUCCESS;
+}
+
+serial::errno
+serial_port_x86::close(void)
+{
+    return serial::SUCCESS;
+}
+
+serial::errno
+serial_port_x86::set_baud_rate(uint32_t baud)
+{
+    if (baud > MAX_BAUD_RATE)
+    {
+        m_baud = MAX_BAUD_RATE;
+    }
+
+    if (baud < MIN_BAUD_RATE)
+    {
+        m_baud = MIN_BAUD_RATE;
+    }
+
+    m_baud = baud;
+
+    // Set the DLAB bit, so we can configure the baud rate
+    // of this port.
+    m_port_io.port_write_8(m_port + FCR_OFFSET, DLAB);
+
+    // Set the baud rate
+    m_port_io.port_write_8(m_port + DL_LSB_OFFSET, baud_to_lo_divisor(m_baud));
+    m_port_io.port_write_8(m_port + DL_MSB_OFFSET, baud_to_hi_divisor(m_baud));
+
+    return serial::SUCCESS;
+}
+
+uint32_t
+serial_port_x86::baud_rate(void)
+{
+    return m_baud;
+}
+
+serial::errno
+serial_port_x86::set_parity_mode(PARITY_MODE parity)
+{
+    uint8_t i = 0;
+    uint8_t data_bits = 0;
+
+    m_parity = parity;
+
+    data_bits = m_port_io.port_read_8(m_port + LCR_OFFSET);
+
+    while (parity_table[i].mode < __PARITY_MAX)
```

Comment:
we will likely need to get rid of the use of __ as it's a pretty big no no with C++ as GCC uses __ for reserved functions.

http://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier